### PR TITLE
Charts 8474 route copy workbook optimize performance requests to postgresql

### PR DESCRIPTION
**Pull request by Sergey-weber**
Optimized the number of queries to the database

Removed the extra child transaction
Made the rights check for the workbook out of the loop
Removed the extra synchronization of links table
Optimize insertGraph entries, set inserting array,  instead of queries in the loop

### CODE REVIEW

**Review by imsitnikov — CHANGES_REQUESTED** [review]

File: src/services/new/workbook/copy-workbook.ts [hunk 1]
```diff
@@ -81,7 +80,9 @@ export const copyWorkbook = async (
 
     const {Workbook} = registry.common.classes.get();
 
-    const originWorkbookModel: Optional<WorkbookModel> = await WorkbookModel.query(getReplica(trx))
+    const targetTrx = getReplica(trx);
+
+    const originWorkbookModel: Optional<WorkbookModel> = await WorkbookModel.query(targetTrx)
```
[the remark this answers is not in the document]
Comment (Sergey-weber) [reply src/services/new/workbook/copy-workbook.ts:85]: the target trx uses getReplace internally by default, I think it's a good practice if we will want to use this method in a transaction

File: src/services/entry/actions/copy-to-workbook.ts [hunk 2]
```diff
@@ -208,27 +145,41 @@ export const copyToWorkbook = async (ctx: CTX, params: Params) => {
...
+    const result = copiedJoinedEntryRevisions.map((entry: {entryId: string | number}) => {
+        return {
+            newJoinedEntryRevision: entry,
+            oldEntryId: mapEntryIdsWithOldIds[entry.entryId],
+        };
     });
 
     logInfo(ctx, BiTrackingLogs.CopyEntry, {
-        entryId: Utils.encodeId(entryId),
+        entryIds: result.map(({newJoinedEntryRevision}) =>
```
[the remark this answers is not in the document]
Comment (Sergey-weber) [reply src/services/entry/actions/copy-to-workbook.ts]: changed on forEach

File: src/services/entry/actions/copy-to-workbook.ts [hunk 3]
```diff
@@ -60,128 +59,66 @@ export const copyToWorkbook = async (ctx: CTX, params: Params) => {
...
-        const [newEntryId, newRevId] = await Promise.all([getId(), getId()]);
-
-        const displayKey = `${newEntryId}/${Utils.getNameByKey({
-            key: originJoinedEntryRevision.displayKey,
-        })}`;
-        const key = displayKey.toLowerCase();
-
-        const links = originJoinedEntryRevision.links as Nullable<Record<string, string>>;
-        if (links) {
-            await Link.sync({entryId: newEntryId, links, ctx, trxOverride: transactionTrx});
```
[the remark this answers is not in the document]
Comment (Sergey-weber) [reply src/services/entry/actions/copy-to-workbook.ts:180]: added skipLinkSync

File: src/services/entry/actions/copy-to-workbook.ts [hunk 4]
```diff
@@ -60,128 +59,66 @@ export const copyToWorkbook = async (ctx: CTX, params: Params) => {
...
-        });
-    }
-
-    if (originJoinedEntryRevision.scope === 'folder') {
-        throw new AppError('Folders cannot be copied', {
-            code: US_ERRORS.FOLDER_COPY_DENIED,
-        });
-    }
-
-    if (!skipWorkbookPermissionsCheck) {
```
[the remark this answers is not in the document]
Comment (Sergey-weber) [reply src/services/entry/actions/copy-to-workbook.ts:102]: fixed

File: src/services/entry/actions/copy-to-workbook.ts [hunk 5]
```diff
@@ -60,128 +59,66 @@ export const copyToWorkbook = async (ctx: CTX, params: Params) => {
...
-        },
-        trx: Entry.replica,
-    });
+        })
+        .whereIn([`${Entry.tableName}.entryId`], entryIds)
+        .timeout(JoinedEntryRevision.DEFAULT_QUERY_TIMEOUT) as unknown as Promise<
+        JoinedEntryRevisionColumns[]
+    >);
+
+    const mapEntryIdsWithOldIds: {[key: string]: string} = {};
```
[the remark this answers is not in the document]
Comment (Sergey-weber) [reply src/services/entry/actions/copy-to-workbook.ts]: fixed

File: src/services/entry/actions/copy-to-workbook.ts [hunk 6]
```diff
@@ -23,10 +28,10 @@ interface Params {
 
 export const validateParams = makeSchemaValidator({
     type: 'object',
-    required: ['entryId', 'destinationWorkbookId'],
+    required: ['entryIds', 'destinationWorkbookId'],
     properties: {
-        entryId: {
-            type: 'string',
+        entryIds: {
+            type: 'array',
```
[the remark this answers is not in the document]
Comment (Sergey-weber) [reply src/services/entry/actions/copy-to-workbook.ts:36]: fixed

File: src/services/new/workbook/copy-workbook.ts [hunk 7]
```diff
@@ -81,7 +80,9 @@ export const copyWorkbook = async (
...
         .select()
         .findById(workbookId)
```
[the remark this answers is not in the document]
Comment (Sergey-weber) [reply src/services/new/workbook/copy-workbook.ts:87]: added

File: src/services/new/workbook/copy-workbook.ts [hunk 8]
```diff
@@ -194,40 +193,52 @@ export const copyWorkbook = async (
             .timeout(WorkbookModel.DEFAULT_QUERY_TIMEOUT);
 
         if (accessServiceEnabled && accessBindingsServiceEnabled) {
```
[the remark this answers is not in the document]
Comment (Sergey-weber) [reply src/services/new/workbook/copy-workbook.ts]: fixed

File: src/services/entry/actions/copy-to-workbook.ts [hunk 9]
```diff
@@ -60,128 +59,66 @@ export const copyToWorkbook = async (ctx: CTX, params: Params) => {
...
-    const {Workbook} = registry.common.classes.get();
-
     const targetTenantId = tenantIdOverride ?? tenantId;
 
-    const originJoinedEntryRevision = await JoinedEntryRevision.findOne({
-        where: {
-            [`${Entry.tableName}.entryId`]: entryId,
+    const transactionTrx = trxOverride ?? Entry.primary;
+
+    const originJoinedEntryRevisions = await (JoinedEntryRevision.query(Entry.replica)
```
[the remark this answers is not in the document]
Comment (Sergey-weber) [reply src/services/entry/actions/copy-to-workbook.ts:69]: we don't have method whereIn in Joined Entry Revision.find, can we provide these in another ticket, we have to understand how we want to realise this method and typings?

[the remark this answers is not in the document]
Comment (Sergey-weber) [reply src/services/entry/actions/copy-to-workbook.ts]: added new method whereIn in JoinedEntryRevision

File: src/services/entry/actions/copy-to-workbook.ts [hunk 10]
```diff
@@ -208,27 +145,41 @@ export const copyToWorkbook = async (ctx: CTX, params: Params) => {
...
-            trx: transactionTrx,
-        });
+            };
+        }),
+    );
 
-        return copiedJoinedEntryRevision;
+    await Entry.query(transactionTrx).insertGraph(newEntries).timeout(Entry.DEFAULT_QUERY_TIMEOUT);
+
+    const copiedJoinedEntryRevisions = await (JoinedEntryRevision.query(transactionTrx)
```
[the remark this answers is not in the document]
Comment (Sergey-weber) [reply src/services/entry/actions/copy-to-workbook.ts]: fixed

**Review by imsitnikov — CHANGES_REQUESTED** [review]

File: src/services/new/workbook/copy-workbook.ts [hunk 11]
```diff
@@ -243,6 +218,24 @@ export const copyWorkbook = async (
             });
         }
 
+        const workbook = new Workbook({
```
[the remark this answers is not in the document]
Comment (Sergey-weber) [reply src/services/new/workbook/copy-workbook.ts]: fixed

File: src/services/entry/actions/copy-to-workbook.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Sergey-weber) [reply src/services/entry/actions/copy-to-workbook.ts:102]: added checking permissions and same workbook id for entries

File: src/services/new/workbook/copy-workbook.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sergey-weber) [reply src/services/new/workbook/copy-workbook.ts]: fixed

File: src/services/entry/actions/copy-to-workbook.ts [hunk 12]
```diff
@@ -60,52 +69,113 @@ export const copyToWorkbook = async (ctx: CTX, params: Params) => {
...
         },
         trx: Entry.replica,
     });
 
-    if (originJoinedEntryRevision === undefined) {
-        throw new AppError('Entry not exists', {
-            code: US_ERRORS.NOT_EXIST_ENTRY,
-        });
-    }
+    const workbookId = originJoinedEntryRevisions[0].workbookId;
```
Comment (imsitnikov) [inline src/services/entry/actions/copy-to-workbook.ts:90]: Some checks aren't needed when using an array of records, and some checks can be moved from an asynchronous loop to a synchronous one. I suggest this:

```ts
if (originJoinedEntryRevisions.length === 0) {
    throw new AppError("Entries don't exist", {
        code: US_ERRORS.NOT_EXIST_ENTRY,
    });
}

let workbookId: Optional<string>;
originJoinedEntryRevisions.forEach((joinedEntryRevision) => {
    if (tenantIdOverride === undefined && joinedEntryRevision.tenantId !== tenantId) {
        throw new AppError(
            `Entry ${Utils.encodeId(joinedEntryRevision.entryId)} doesn't exist`,
            {
                code: US_ERRORS.NOT_EXIST_ENTRY,
            },
        );
    }

    if (joinedEntryRevision.scope === 'folder') {
        throw new AppError('Folders cannot be copied', {
            code: US_ERRORS.FOLDER_COPY_DENIED,
        });
    }

    if (joinedEntryRevision.workbookId === null) {
        throw new AppError(
            `Entry ${Utils.encodeId(
                joinedEntryRevision.entryId,
            )} doesn't have a workbookId and cannot be copied to a workbook.`,
            {
                code: US_ERRORS.ENTRY_WITHOUT_WORKBOOK_ID_COPY_DENIED,
            },
        );
    }

    if (workbookId === undefined) {
        workbookId = joinedEntryRevision.workbookId;
    } else if (joinedEntryRevision.workbookId !== workbookId) {
        throw new AppError(
            `Copying entries from different workbooks is denied – ${Utils.encodeId(
                workbookId,
            )} and ${Utils.encodeId(joinedEntryRevision.workbookId)}`,
            {
                code: US_ERRORS.ENTRIES_WITH_DIFFERENT_WORKBOOK_IDS_COPY_DENIED, // <-- new error code
            },
        );
    }
});

if (workbookId === undefined) {
    throw new AppError(`Entries don't have a workbookId and cannot be copied to a workbook.`, {
        code: US_ERRORS.ENTRY_WITHOUT_WORKBOOK_ID_COPY_DENIED,
    });
}

const mapEntryIdsWithOldIds = new Map<string, string>();

const newEntries = await Promise.all(
    originJoinedEntryRevisions.map(async (originJoinedEntryRevision) => {
        const [newEntryId, newRevId] = await Promise.all([getId(), getId()]);

        ...
    }),
);

if (!skipWorkbookPermissionsCheck) { // <-- no need to check workbookId
	...
}
```

**Review by imsitnikov — CHANGES_REQUESTED** [review]

File: src/services/entry/actions/copy-to-workbook.ts (hunk 12 above)
Comment (Sergey-weber) [reply src/services/entry/actions/copy-to-workbook.ts]: yea, fixed it

**Review by imsitnikov — APPROVED** [review]

**Pull request merged**